Integrate feedback distribution within nbgrader

danielmaitre opened the pull request:
This is the result of the work of @dsblank, @perllaghu, @billywardrop29 and @danielmaitre at the Edinburgh hackathon.  

It includes UI integration of the feedback step, release of the feedback to students in the exchange, and fetching of this feedback by the students.

Some UI work needs to be finished but starting a PR will help check the coverage of the existing tests.

Fixes #795
Replaces #1096

jhamrick commented:
Awesome, I will review this on the train 🙂 

Looks like there's a merge conflict since I merged #1119, so I think this will need to be rebased unfortunately.

dsblank commented:
@danielmaitre I made a PR on yours that is a few steps closer on the submission display. I think there is a bug somewhere in fetch and display with proper available/view links. But I ran out of time today.

I can help refine all of this tomorrow on the plane home (I have a good 7 hours :) But maybe you will have solved all of the issues :)

jhamrick commented on nbgrader/exchange/releaseFeedback.py:
```diff
@@ -0,0 +1,67 @@
+import os
```
Maybe this file could be called `release_feedback.py`, to be consistent with the name of the nbgrader command?

danielmaitre replied:
yes

jhamrick commented on nbgrader/exchange/fetchfeedback.py:
```diff
@@ -0,0 +1,66 @@
+import os
```
Maybe this file could be called `fetch_feedback.py`, to be consistent with the name of the nbgrader command?

danielmaitre replied:
yes

jhamrick commented on nbgrader/exchange/releaseFeedback.py:
```diff
@@ -0,0 +1,67 @@
+            self.outbound_feedback_path,
+            S_IRUSR|S_IWUSR|S_IXUSR|S_IXGRP|S_IXOTH
+        )
+    # should be moved up to Exchange? Common to this and release
```
That would be great :)

danielmaitre replied:
Will do.

jhamrick commented:
P.S. I also think the docs need to be regenerated after the rebase, it looks like a bunch of the docs got removed so I think if you do a clean regeneration and commit the results the diff should be a bit smaller.

danielmaitre commented on nbgrader/apps/api.py:
```diff
@@ -979,3 +979,104 @@ def autograde(self, assignment_id, student_id, force=True, create=True):
+            ret_dic = capture_log(app)
+            # assignment tab needs a 'value' field with  
+        with temp_attrs(self.coursedir, assignment_id='*', student_id=student_id):
+            lister_rel = ExchangeList(inbound=False, coursedir=self.coursedir, config=self.config)
```
The buttons on the assignment_list tab require the POST to return an updated list of assignments. lister_sub and lister_rel seemed to return complementary lists to fill the two blocks on the assignment_list tab. When I left one out only one of the two blocks would fill.

jhamrick replied:
Ah ok yeah that makes sense, thanks! Maybe you could leave a comment in the code just explaining that?

dsblank replied:
I have a better version of this code in my PR.

danielmaitre replied:
I'll add one.

danielmaitre commented on nbgrader/apps/api.py:
```diff
@@ -979,3 +979,104 @@ def autograde(self, assignment_id, student_id, force=True, create=True):
+                            student_id=student_id):
+                app = Feedback(coursedir=self.coursedir, parent=self)
+                c = Config()
+                c.HTMLExporter.template_file = 'feedback.tpl'
```
@dsblank should be able to comment on this one.

dsblank replied:
The issue here was that the HTMLExporter was being used in other place(s) in conversion. The code in the Feedback converter would only set the template to "feedback.tpl" if the template had not  already been set in the config. But it had been set, so we have to manually set it here to override that. (Hmmm, I wonder if we need to set it back...?)

jhamrick replied:
Thanks for clarifying---can you add a comment in the code with this explanation?

> I wonder if we need to set it back...?

I think probably not, since it creates a new config object on the fly here.

dsblank replied:
Added comments and refactor in https://github.com/danielmaitre/nbgrader/pull/8

danielmaitre commented:
Concerning the two identical submissions submitted twice but at different times, I personally prefer that they have the same exact feedback. The duplication happens often with my students when they are not sure the submission worked and they submit 5 times just to be sure. Probably the better solution would be to reject a submission (very politely) if it is identical to the previously submitted submission, this way we would get rid of virtually all duplicates?
--


"

jhamrick commented:
> the better solution would be to reject a submission (very politely) if it is identical to the previously submitted submission, this way we would get rid of virtually all duplicates?

I think that sounds like a good solution (though it would make testing a bit more annoying as it would require constantly clearing the exchange :) )

jhamrick commented on nbgrader/apps/fetchfeedbackapp.py:
```diff
@@ -0,0 +1,95 @@
+        return classes
+
+    def _load_config(self, cfg, **kwargs):
+        if 'FetchApp' in cfg:
```
I don't think you need this here, as this is a deprecation warning for the regular (assignment) fetch functionality.

dsblank replied:
Removed method in https://github.com/danielmaitre/nbgrader/pull/8

jhamrick commented on nbgrader/apps/releasefeedbackapp.py:
```diff
@@ -0,0 +1,81 @@
+        return classes
+
+    def _load_config(self, cfg, **kwargs):
+        if 'ReleaseFeedbackApp' in cfg:
```
Same thing here, I don't think you need this warning.

dsblank replied:
Removed method in https://github.com/danielmaitre/nbgrader/pull/8

jhamrick commented on nbgrader/exchange/fetch_feedback.py:
```diff
@@ -0,0 +1,66 @@
+            newName = noExt + '.html'
+            htmlFile = os.path.join(destWithTimestamp, newName)
+            shutil.copy(feedbackpath, htmlFile)
+        #else:
```
Can you remove this commented code?

dsblank replied:
Done in https://github.com/danielmaitre/nbgrader/pull/8

jhamrick commented on nbgrader/nbextensions/assignment_list/assignment_list.js:
```diff
@@ -225,24 +225,92 @@ define([
         for (var i=0; i<len; i++) {
             var element = $('<div/>');
-            var item = new Assignment(element, data[i], this.fetched_selector, $.proxy(this.handle_load_list, this), this.options);
+            var item = new Assignment(element, submissions, data[i], this.fetched_selector,
```
Ah, I didn't realize it at first but I guess class populates `submissions`? That's a bit non-obvious---maybe leave a comment in the code saying that's what's happening.

Actually even better, if you don't mind making a few changes, would be not to pass `submissions` into the constructor but create it within the `Assignment` class and then below when you iterate over the submissions read it out as an attribute of the class.

dsblank replied:
Done in https://github.com/danielmaitre/nbgrader/pull/8

jhamrick commented on nbgrader/docs/source/command_line_tools/index.rst:
```diff
@@ -21,7 +21,7 @@ Instructor commands
     nbgrader-update
     nbgrader-autograde
-    nbgrader-feedback
+    nbgrader-generate-feedback
```
Is this correct? It doesn't look like you've renamed the command line app to `nbgrader generate_feedback` as far as I can tell---isn't it still `nbgrader feedback`?

dsblank replied:
Done

jhamrick commented on nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb:
```diff
@@ -1059,8 +1059,11 @@
     "\n",
-    "    :doc:`/command_line_tools/nbgrader-feedback`\n",
-    "        Command line options for ``nbgrader feedback``\n",
+    "    :doc:`/command_line_tools/nbgrader-generate-feedback`\n",
```
Same comment here about `nbgrader feedback`

dsblank replied:
Done

jhamrick commented on nbgrader/apps/api.py:
```diff
@@ -981,3 +981,104 @@ def autograde(self, assignment_id, student_id, force=True, create=True):
             return capture_log(app)
+
+    def generate_feedback(self, assignment_id, student_id=None):
+        """Run ``nbgrader generate_feedback`` for a particular assignment and student.
```
```suggestion
        """Run ``nbgrader feedback`` for a particular assignment and student.
```

dsblank replied:
Done

jhamrick commented on nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb:
```diff
@@ -1084,13 +1087,29 @@
+    "\n",
+    "Generating feedback will create HTML files in the local instructor directory. Releasing feedback will copy those HTML files to the nbgrader exchange.\n",
+    "\n",
+    "We can generate feedback based on the graded notebooks by running the `nbgrader generate_feedback` command, which will produce HTML versions of these notebooks at the following location:\n",
```
```suggestion
    "We can generate feedback based on the graded notebooks by running the `nbgrader feedback` command, which will produce HTML versions of these notebooks at the following location:\n",
```

dsblank replied:
Done

jhamrick commented on nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb:
````diff
@@ -1084,13 +1087,29 @@
     "```\n",
     "\n",
-    "Unfortunately, `nbgrader feedback` is not yet available through the formgrader interface."
+    "The `nbgrader generate_feedback` is available by clicking the Generate Feedback button on either the Manage Assignments view (to generate feedback for all graded submissions), or on the individual student's Manage Submission page (to generate feedback for that specific individual)."
````
```suggestion
    "The `nbgrader feedback` command is available by clicking the Generate Feedback button on either the Manage Assignments view (to generate feedback for all graded submissions), or on the individual student's Manage Submission page (to generate feedback for that specific individual)."
```

dsblank replied:
Done

jhamrick commented on nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb:
```diff
@@ -1129,32 +1150,99 @@
     "%%bash\n",
     "\n",
-    "nbgrader feedback \"ps1\" --force"
+    "nbgrader generate_feedback \"ps1\" "
```
```suggestion
    "nbgrader feedback \"ps1\" "
```

dsblank replied:
Done

jhamrick commented on nbgrader/docs/source/user_guide/highlights.rst:
```diff
@@ -42,3 +43,11 @@ commands:
 * `nbgrader release`: release a notebook to students
 * `nbgrader collect`: collect students' submissions
 * `nbgrader autograde`: autograde students' submissions
+* `nbgrader generate_feedback`: create feedback files from graded submissions
```
```suggestion
* `nbgrader feedback`: create feedback files from graded submissions
```

dsblank replied:
Done

dsblank commented:
@danielmaitre Unless you have already started on it, I'll make the submissions changes now as requested from Jess. After that, I'll look at the rest unless I hear from you. I have a few hours on the plane now. :)

dsblank commented:
@danielmaitre Added the submissions changes as a PR on your repo: https://github.com/danielmaitre/nbgrader/pull/8

dsblank commented:
@jhamrick @danielmaitre I intended for there to be three `nbgrader` commands:

* `nbgrader generate_feedback` (for instructor with exchange)
* `nbgrader release_feedback` (for instructor with exchange)
* `nbgrader fetch_feedback` (for student with exchange)

I think it would be confusing to call one of these `feedback`.

dsblank commented:
(If we need to create an alias to `feedback` which we can deprecate one day, please let me know).

danielmaitre commented:
HI @dsblank , I won't look at this until tomorrow morning 6h30 am, so please make all the changes you want.
We broke the nbgrader/tests/nbextensions/test_assignment_list.py test. I think the problem is that none of the elements in the reworked submitted assignment list has a "list_item" class. We seem to have lost it on the way and it also causes the formatting to go weird.
Have a good flight!

dsblank commented:
@danielmaitre Ok, sounds like a plan! I'll look at the bug above, and continue to make incremental fixes to the PR. The UI now looks like: 

![Screenshot from 2019-06-02 09-04-26](https://user-images.githubusercontent.com/168568/58761790-d83a0800-8515-11e9-9cfd-3aa602ee7559.png)

which is similar to our original goals. I'm not sure I'll be able to do the expand/collapse because of updating the view when we click "Fetch Feedback"... exploring...

jhamrick commented:
Cool, that sounds good! Re: `feedback` vs. `generate_feedback`, I am fine if it is `generate_feedback` but then there need to be a bunch of changes made as well to rename feedback to generate feedback. What might be easiest for now is to just keep the name `feedback` and in a separate PR rename `feedback` to `generate_feedback`, so this one doesn't get too big.

The screenshot looks nice! What does "available" mean though? That feedback is available to be fetched?

jhamrick commented:
Another couple things (sorry I keep thinking of stuff)

1. You probably want to mark the tests you've added as `@notwindows` because the exchange functionality is not supported on windows.
2. You probably want to disable the "Release feedback" buttons in the UI if the exchange isn't available (which will be the case either on windows or when instructors haven't set up the exchange properly)

dsblank commented:
Thanks @danielmaitre for fixing the tests to match the new submission UI.

Things left to do before merging:

- [ ] resolve conflicts
- [ ] mark tests notwindows
- [ ] hide ui parts when no exchange available
- [ ] change "available" in submission ui listing to "view" link
- [ ] don't change nbgrader command "release" to "generate_release" yet

I'll start on these (in no particular order) now unless I hear otherwise. 

Almost finished with this PR!

dsblank commented on nbgrader/exchange/list.py:
```diff
@@ -79,20 +79,36 @@ def parse_assignments(self):
+                _, notebookFilename = os.path.split(notebook)
+                notebookName, _ = os.path.splitext(notebookFilename)
+                feedbackpath = os.path.join(self.root, info['course_id'], 'feedback', '{0}.html'.format(nb_hash))  
+                localFeedbackpath = os.path.join(self.coursedir.root, info['course_id'], 'feedback', '{0}.html'.format(notebookName))  
```
@danielmaitre I believe that the path assumes that the files are `course_id/feedback` but that should be either:

1. `course_id/assignment_id/feedback` (if organized by course), or
2. `assignment_id/feedback` (if not organized by course)

perllaghu replied:
The current `fetch.py` code in the exchange does:

```python
        if self.path_includes_course:
            root = os.path.join(self.course_id, self.coursedir.assignment_id)
        else:
            root = self.coursedir.assignment_id
        self.dest_path = os.path.abspath(os.path.join('.', root))
```

danielmaitre replied:
I'll fix the merge confilcts

danielmaitre replied:
Yes the local path was not set properly, I wrote this bit before we decided what directory structure we wanted, and forgot to adapt it. I've now changed it and added the timestamp as well so now the UI shows "view" instead of available when the file is present.

danielmaitre replied:
fixed a few more missing @notwindows

dsblank replied:
Looks good! I'm turning the "view" into a link, and the "available" to "ready to fetch". Will make a PR within the next hour or so.

dsblank replied:
PR adds link to "view", and changes text to "ready to fetch": https://github.com/danielmaitre/nbgrader/pull/9

dsblank commented:
Here is the current state of the Submissions UI:

![Screenshot from 2019-06-03 09-01-26](https://user-images.githubusercontent.com/168568/58803786-477b2f00-85de-11e9-9571-d16475c19deb.png)

I guess we should remove the `Fetch Feedback` button if feedback has already been fetched for all submissions for that assignment. I originally left it there so students could easily re-fetch any feedback that had been re-released. But that is not consistent with the fetch and release of assignments.

perllaghu commented:
Yep, remove it....
And Yes, once an assignment has been fetched [and exists on disk] it cannot
be re-fetched/updated

There is an argument that if there's feedback [for a submission] that's not
yet been fetched, and will therefore go into a directory that doesn't yet
exist, then there should be a Fetch button to get it

On Mon, 3 Jun 2019 at 14:04, Doug Blank <notifications@github.com> wrote:

> I guess we should remove the Fetch Feedback button if feedback has
> already been fetched for all submissions for that assignment. I originally
> left it there so students could easily re-fetch any feedback that had been
> re-released. But that is not consistent with the fetch and release of
> assignments.
>

-- 
Ian

danielmaitre commented:
I've added an entry "allFeedbackDownloaded" in the assignment info passed to the UI along with "hasFeedback", and added logic to the UI to remove the button when there is nothing to download. Before this commit the CI tests were working, finger crossed they still will...

dsblank commented:
I'll update the UI based on the new entry, and tie up the remaining loose ends in about 8 hours.

perllaghu commented:
With the current exchange, I think you need to - when the exchange has a
proper API, then that logic [_allDownloaded_] should be moved into the
server_extension code

On Mon, 3 Jun 2019 at 15:36, danielmaitre <notifications@github.com> wrote:

> I've added an entry "allFeedbackDownloaded" in the assignment info passed
> to the UI along with "hasFeedback", and added logic to the UI to remove the
> button when there is nothing to download. Before this commit the CI tests
> were working, finger crossed they still will...
>

-- 
Ian

jhamrick commented:
This is coming along really nicely 😄Let me know when you're ready for me to do a final pass!

dsblank commented:
@danielmaitre One more PR: https://github.com/danielmaitre/nbgrader/pull/10

That addresses everything that I know about, I believe. Woot!

There is the artifact that if you submit an assignment more than once, then the feedback for one will be the feedback for all (and you'll see the "Fetch Feedback" button appear, and clicking it will download the old feedback for that one too). But I think we can address that by preventing submitting a notebook twice. In any event, we can refine that process after this PR.

@jhamrick Of course, after Daniel pulls 10, please let us know if there is anything you'd like changed.

danielmaitre commented:
There are problems with the automatic CI tests, on travis only with python nighty and on appveyor only for python 3.5/6 (not sure which one it is, the executable says 3.5, the environment 3.6), anyone has any idea how to fix this?

perllaghu commented:
@BertR - you did some stuff with the Travis at the hackathon, got time to have a look?

dsblank commented:
Both errors on AppVeyor and Travis look like issues in nbconvert that has a missing language_info key in something. I don't think that is related to anything we (or nbgrader) has changed.

jhamrick commented:
> Both errors on AppVeyor and Travis look like issues in nbconvert that has a missing language_info key in something. I don't think that is related to anything we (or nbgrader) has changed.

Yes, that is my suspicion as well, I have restarted the tests to see.

jhamrick commented on .gitignore:
```diff
@@ -112,3 +112,6 @@ package-lock.json
 .goutputstream*
 
 
+fakecoursedir/
```
Are these changes needed for this PR? They don't seem to actually appear in the other changes.

dsblank replied:
Removed.

jhamrick commented on nbgrader/nbextensions/assignment_list/assignment_list.js:
```diff
@@ -315,14 +389,51 @@ define([
     Assignment.prototype.make_row = function () {
         var row = $('<div/>').addClass('col-md-12');
-        row.append(this.make_link());
+	var link = this.make_link();
```
```suggestion
	 var link = this.make_link();
```

dsblank replied:
Done

jhamrick commented on nbgrader/server_extensions/formgrader/apihandlers.py:
```diff
@@ -272,13 +273,54 @@ def post(self, assignment_id, student_id):
         self.write(json.dumps(self.api.autograde(assignment_id, student_id)))
 
 
+class GenerateAllFeedbackHandler(BaseApiHandler):
```
Can you add the `@check_notebook_dir` decorator to all these new routes? (Sorry, it's from a PR I just merged recently)

dsblank replied:
Done

dsblank commented:
@jhamrick @danielmaitre Final fixes can be found in https://github.com/danielmaitre/nbgrader/pull/11

danielmaitre commented:
All checks have passed!

perllaghu commented:
WOO-HOO!!!

Dang - that was some good work guys...

dsblank commented:
I very much enjoyed working with you all! Great project to dive into most of the topics of nbgrader, and a great team! Let's do it again :)

jhamrick commented:
Fabulous, thanks so much to all of you for working on this!

dsblank commented:
Party!

Thanks for being our mentor through this @jhamrick ! I know how much patience it takes. Much appreciated!

jhamrick commented:
I am more than happy to---it's really exciting seeing other want to contribute! 😄

danielmaitre commented:
Thank you for your help on this project! 
Github says I can safely delete the branch, should I?

jhamrick commented:
Fine by me.

rkdarst commented:
I was looking at this, wanting to take it into our use.  It would be very nice to have and save us a lot of work!  But first I had to figure out how it works.  I was hoping that someone could confirm my observations below:

First, it seems to list the exchange inbound path (submitted assignments) to get notebook hashes, which are used to copy assignments back.  But the security of the inbound path is by not allowing students to list it.  Is this the case, do students have to list the inbound directory in order to fetch feedback?

Second, it also ensures that `{course_directory}/feedback/` is read/execute by students and fails with an error if not.  This would mean that students can list that entire directory and see all other students feedbacks.  Is this the expected situation?  At first I thought this could make the feedback dir read but not execute, so that students could get their feedback if they knew their exact filename (based on the hash), but not list all other assignments.

My thought, if the above is the case, chaneg it to do this: find the the hashes of the submitted notebooks from `~/.local/share/jupyter/nbgrader_cache/{course_id}/`, which only the submitter would have for their own notebooks.  Then, use that to get the feedback from the feedback dir, which they can access to copy known filenames from but not list.

jhamrick commented:
I am not sure the answers to your questions off the top of my head, but can help try to answer them later this weekend 😄 in the meantime perhaps @danielmaitre @dsblank @perllaghu would know?

Sefriol commented:
I mimicked how [Feedback Exchanger](https://github.com/jupyter/nbgrader/blob/master/nbgrader/exchange/fetch_feedback.py#L26) does determine submission and it's feedback in our environment.

By testing this with couple of submissions, you can indeed generate the same feedback md5-hash from `~/.local/share/jupyter/nbgrader_cache/{course_id}/` as you do from inbound directory which would guide the student to right feedback form. With this method student could then copy the feedback without giving student `ls` rights to the inbound directory. Much preferred way of doing things.

jhamrick commented:
Let's continue discussion in #1202
